crates/amaru-ledger/src/rules/transaction/collateral.rs

```diff
@@ -0,0 +1,363 @@
+// Copyright 2025 PRAGMA
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    collections::BTreeMap,
+    fmt::{self},
+};
+
+use amaru_kernel::{
+    AlonzoValue, MemoizedTransactionOutput, MintedTransactionOutput, TransactionInput,
+    TransactionInputAdapter, Value, protocol_parameters::ProtocolParameters,
+};
+use thiserror::Error;
+
+use crate::context::UtxoSlice;
+
+/*
+ * CollateralBalance is used to track difference in collateral input vlaue and collateral return value.
+ * The value of everything should be zero in this struct, otherwise value is not conserved.
+ * We allow negative values here so that we are able to display them in an error message
+ *
+ * i64 has a lower maximum than u64 due to the data structure, but the value for coins must always be in the range:
+ *  [-9223372036854775808, 18446744073709551615]
+ * Which is within the bounds of an i64, so conversions are safe i64<->u64
+ */
+pub struct CollateralBalance {
+    pub coin: i64,
+    pub multiasset: BTreeMap<Vec<u8>, i64>,
+}
+
+impl fmt::Debug for CollateralBalance {
```

Comment:
```suggestion
impl fmt::Display for CollateralBalance {
```

Do not write `Debug` instances by hand, unless you have a really good reason to.